suse/factory.go

```diff
@@ -0,0 +1,233 @@
+package suse
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strings"
+	"sync"
+
+	"github.com/quay/claircore"
+	"github.com/quay/claircore/libvuln/driver"
+	"github.com/quay/claircore/pkg/ovalutil"
+	"github.com/quay/zlog"
+	"golang.org/x/net/html"
+)
+
+//doc:url updater
+const base = `https://ftp.suse.com/pub/projects/security/oval/`
+
+var (
+	reELFile   = regexp.MustCompile(`suse.linux.enterprise.server.([0-9]+).xml.bz2`)
+	reLeapFile = regexp.MustCompile(`opensuse.leap.([0-9]+.[0-9]+).xml.bz2`)
+)
+
+type Factory struct {
+	c    *http.Client
+	base *url.URL
+
+	mu   sync.Mutex
+	etag string
+	cur  driver.UpdaterSet
+}
+
+// UpdaterSet implements [driver.UpdaterSetFactory].
+func (f *Factory) UpdaterSet(ctx context.Context) (driver.UpdaterSet, error) {
+	ctx = zlog.ContextWithValues(ctx, "component", "suse/Factory.UpdaterSet")
+	us := driver.NewUpdaterSet()
+	if f.c == nil {
+		zlog.Info(ctx).
+			Msg("unconfigured")
+		return us, nil
+	}
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, f.base.String(), nil)
+	if err != nil {
+		return us, fmt.Errorf("suse: unable to construct request: %w", err)
+	}
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	if f.etag != "" {
+		req.Header.Set(`if-none-match`, f.etag)
+	}
+	zlog.Debug(ctx).
+		Stringer("url", f.base).
+		Msg("making request")
+	res, err := f.c.Do(req)
+	if err != nil {
+		return us, fmt.Errorf("suse: error requesting %q: %w", f.base.String(), err)
+	}
+	defer res.Body.Close()
+	switch res.StatusCode {
+	case http.StatusNotModified:
+		zlog.Debug(ctx).
+			Stringer("url", f.base).
+			Msg("not modified")
+		return f.cur, nil
+	case http.StatusOK:
+	default:
+		return us, fmt.Errorf("suse: unexpected status requesting OVAL dir: %v", res.Status)
+	}
+
+	dir, err := html.Parse(res.Body)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// It's kind of gross having to parse the HTML but the other path,
```

Comment:
I am pretty thankful for https://security.access.redhat.com/data/oval/v2/PULP_MANIFEST when I look at the alternative below 😂